File: M_Simulator_FEM_2D/08_M_Simulator_FEM_2D/MATLAB/src/Global_Stiffness_Builder.m
function [k_global] = Global_Stiffness_Builder(Gen)

%Predefine memory of global stiffness matrix
k_global = zeros(2*Gen.Nn,2*Gen.Nn);                        %[2*Nn,2*Nn]

%Build global stiffness matrix
for i = 1:Gen.Ne
    %Find nodes in element
    Nodes = find(Gen.Ref(i,:),4);                           %[1,4]
    %Prepare local stiffness matrix of element
    %Plane strain
    [kl] = Local_Stiffness_PStrain(Gen);              %[8,8]
    %Plane stress
    %[kl] = Local_Stiffness_PStress(Gen);              %[8,8] 
    
    %Add element stiffness matrix to global stiffness matrix
    %Apply local stiffness matrix contribution to each node of the element
    for j = 1:8
        row = Nodes(floor((j-1)/2) + 1);                    %[1,1]
        loc = 2*(row) - 1 + rem(j-1,2);                     %[1,1]
        
        %Pick all even and odd columns of row of k_global
        odd = kl(j,1:2:length(kl(j,:)));                    %[1,4]
        even = kl(j,2:2:length(kl(j,:)));                   %[1,4]
        
        %Find the columns they correspond to
        oddI = Nodes*2 - 1;                                 %[1,4]
        evenI = Nodes*2;                                    %[1,4]
        
        %Add local stiffness matrix to global stiffness matrix
        k_global(loc,oddI) = k_global(loc,oddI) + odd;      %[2*Nn,2*Nn]
        k_global(loc,evenI) = k_global(loc,evenI) + even;   %[2*Nn,2*Nn]
        
    end 
end

end

File: HM_Simulator_FEM_FVM_2D/09_HM_Simulator_FEM_FVM_2D/Python/src/src_mech/global_stiffness_builder.py
import numpy as np
from src.src_mech.local_stiffness_pstrain import local_stiffness_pstrain
from src.src_mech.local_stiffness_pstress import local_stiffness_pstress

def global_stiffness_builder(Gen):

    #Predefine memory of global stiffness matrix
    k_global = np.zeros((2*Gen["Nn"],2*Gen["Nn"]))

    #Catch ill-defined plane strain or stress
    if Gen["plane"] not in ["strain", "stress"]:
        raise ValueError("Gen['plane'] must be 'strain' or 'stress'")
    #Prepare local stiffness matrix of element (only need to do once since elements same size)
    if Gen["plane"] == "strain":
        #For plane strain conditions
        kl = local_stiffness_pstrain(Gen)
    else:
        #For plane stress conditions
        kl = local_stiffness_pstress(Gen)

    #Build global stiffness matrix
    for i in range(Gen["Ne"]):
        #Find nodes in element
        nodes = Gen["Ref"][i]

        # Global degree-of-freedom indices for this element
        # Ordering: [u1, v1, u2, v2, u3, v3, u4, v4]
        dof = np.array([
            2 * nodes[0], 2 * nodes[0] + 1,
            2 * nodes[1], 2 * nodes[1] + 1,
            2 * nodes[2], 2 * nodes[2] + 1,
            2 * nodes[3], 2 * nodes[3] + 1
        ])

        # Add local stiffness contributions to the global DOFs associated with this element
        k_global[np.ix_(dof, dof)] += kl

    return k_global